### PR TITLE
[COREV] Define ALU instructions and add tests

### PR DESCRIPTION
**Pull request by serkm**

### CODE REVIEW

**Comment by flip1995** [comment]
Thanks for the PRs! I probably have time to review those next week. In the meantime, can you resolve the merge conflicts please?

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 1]
```diff
@@ -176,6 +177,178 @@ let Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore =
                     Sched<[]>;
 } // Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore = 0
 
+let Predicates = [HasExtXCoreVAlu], hasSideEffects = 0, mayLoad = 0, mayStore = 0 in {
+
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]:
```suggestion
```

File: llvm/lib/Target/RISCV/RISCVInstrFormatsCOREV.td [hunk 2]
```diff
@@ -113,3 +113,99 @@ class RVInstMac16I<bits<2> funct2, bits<3> funct3, dag outs, dag ins,
   let Inst{11-7} = rd;
   let Opcode = OPC_CUSTOM2.Value;
 }
+
+class RVInstAlu_rii<bits<2> funct2, bits<3> funct3, dag outs, dag ins,
+                    string opcodestr, string argstr, list<dag> pattern>
+    : RVInst<outs, ins, opcodestr, argstr, pattern, InstFormatOther> {
+  bits<5> Is3;
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrFormatsCOREV.td]: Immediate are usually called `imm<#bits>`, so this would be called
```suggestion
  bits<5> imm5;
```

Since there is no reference on how to name immediate, if there are two with the same size, I suggest to use `imm5_1` `imm5_2` here.

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 3]
```diff
@@ -176,6 +177,178 @@ let Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore =
...
+  // Bit manipulation operations
+  def CV_EXTRACT   : RVInstAlu_rii<0b11, 0b000, (outs GPR:$rd), (ins GPR:$rs1, uimm5:$Is3, uimm5:$Is2),
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]: Also here (and everywhere below), the immediate naming convention should be applied
```suggestion
  def CV_EXTRACT   : RVInstAlu_rii<0b11, 0b000, (outs GPR:$rd), (ins GPR:$rs1, uimm5:$imm5, uimm5:$imm5),
```

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 4]
```diff
@@ -176,6 +177,178 @@ let Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore =
...
+                     Sched<[]>;
+  def CV_BCLRR     : RVInstAlu_rr<0b1000000, 0b011, (outs GPR:$rd), (ins GPR:$rs1, GPR:$rs2),
+                                  "cv.bclrr", "$rd, $rs1, $rs2", []>,
+                     Sched<[]>;
+  def CV_BSETR     : RVInstAlu_rr<0b1000000, 0b100, (outs GPR:$rd), (ins GPR:$rs1, GPR:$rs2),
+                                  "cv.bsetr", "$rd, $rs1, $rs2", []>,
+                     Sched<[]>;
+  def CV_BITREV    : RVInstAlu_rii<0b11, 0b101, (outs GPR:$rd), (ins GPR:$rs1, uimm5:$Is3, uimm5:$Is2),
+                                   "cv.bitrev", "$rd, $rs1, $Is3, $Is2", []>,
+                     Sched<[]>;
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]: This instruction only takes a `uimm2` as the third argument. But then again, the first 3 bits just get ignored, if a `uimm5` is passed.

Maybe create a special operand type for this? I'm not sure if that is necessary though.

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 5]
```diff
@@ -176,6 +177,178 @@ let Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore =
...
+
+  let DecoderNamespace = "COREV_ALU" in
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]: Why is this necessary for the `cv.ror` instruction?

File: llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp [hunk 6]
```diff
@@ -367,6 +367,18 @@ DecodeStatus RISCVDisassembler::getInstruction(MCInst &MI, uint64_t &Size,
...
+      LLVM_DEBUG(
+          dbgs() << "Trying COREV_ALU32 table (32-bit Instruction):\n");
+      // Calling the auto-generated decoder function.
+      Result = decodeInstruction(DecoderTableCOREV_ALU32, MI, Insn, Address,
+                                 this, STI);
+      if (Result != MCDisassembler::Fail) {
+        Size = 4;
+        return Result;
+      }
+    }
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp]: IIUC this only deals with the `cv.ror` command? I don't quite get, why this is necessary. AFAICT no CORE-V instructions can even reach this code, since all of the instructions have an opcode ending in `0b11`, since cv32e40p is a 32-bit platform.

**Review by flip1995 — CHANGES_REQUESTED** [review]
IIUC the [bitmanip extension is not implemented by the CORE-V GCC compiler](https://github.com/openhwgroup/core-v-docs/pull/269#discussion_r528748212), since it is superseded by the RISC-V B extension (once this becomes ratified). I'm not sure if we should include those here. cc @jeremybennett 

Tests LGTM

**Comment by jeremybennett** [comment]
> IIUC the [bitmanip extension is not implemented by the CORE-V GCC compiler](https://github.com/openhwgroup/core-v-docs/pull/269#discussion_r528748212), since it is superseded by the RISC-V B extension (once this becomes ratified). I'm not sure if we should include those here. cc @jeremybennett
> 
> Tests LGTM

@serka

> IIUC the [bitmanip extension is not implemented by the CORE-V GCC compiler](https://github.com/openhwgroup/core-v-docs/pull/269#discussion_r528748212), since it is superseded by the RISC-V B extension (once this becomes ratified). I'm not sure if we should include those here. cc @jeremybennett
> 
> Tests LGTM

@SerkanMuhcu @flip1995 You are correct that we are not implementing the PULP bit manip extension in the CORE-V GCC. It is not implemented n the standard CV32E40P and will be replaced by the standard B extension in the V2 chip.

**Comment by serkm** [comment]
Ok, I'll remove the bitmanip instructions.

@jeremybennett Did I understand correctly that the SIMD instructions are not supported aswell?

File: llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp (hunk 6 above)
Comment (serkm) [reply llvm/lib/Target/RISCV/Disassembler/RISCVDisassembler.cpp]: Yes, I must have made a mistake here.

The different decoding namespace is necessary, because otherwise the disassembler creates an `<unknown>` when there is an ambiguous instruction encoding, making some tests on the disassembler fail.

Anyway, this shouldn't matter, since `cv.ror` isn't supported anymore.

**Comment by flip1995** [comment]
Yes, the SIMD instructions are superseded by the RISC-V P extension.

File: llvm/lib/Target/RISCV/RISCVInstrFormatsCOREV.td [hunk 7]
```diff
@@ -113,3 +113,82 @@ class RVInstMac16I<bits<2> funct2, bits<3> funct3, dag outs, dag ins,
...
+  bits<5> Imm5;
+  bits<5> rs1;
+
+  let Inst{31} = Imm12{11};
+  let Inst{30-25} = Imm12{9-4};
+  let Inst{24-20} = Imm5;
+  let Inst{19-15} = rs1;
+  let Inst{14-12} = funct3;
+  let Inst{11-8} = Imm12{3-0};
+  let Inst{7} = Imm12{10};
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrFormatsCOREV.td]:
```suggestion
  bits<12> imm12;
  bits<5> imm5;
  bits<5> rs1;

  let Inst{31} = imm12{11};
  let Inst{30-25} = imm12{9-4};
  let Inst{24-20} = imm5;
  let Inst{19-15} = rs1;
  let Inst{14-12} = funct3;
  let Inst{11-8} = imm12{3-0};
  let Inst{7} = imm12{10};
```

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 8]
```diff
@@ -176,6 +177,121 @@ let Predicates = [HasExtXCoreVMac], hasSideEffects = 0, mayLoad = 0, mayStore =
...
+} // Predicates = [HasExtXCoreVAlu], hasSideEffects = 0, mayLoad = 0, mayStore = 0, Opcode = OPC_CUSTOM2.Value, Constraints = "$rd = $rd_wb"
+
+let Predicates = [HasExtXCoreVAlu], hasSideEffects = 0, mayLoad = 0, mayStore = 0, isBranch = 1, isTerminator = 1 in {
+  // Immediate branching operations
+  def CV_BEQIMM : RVInstImmBranch<0b010, (outs), (ins GPR:$rs1, simm5:$Imm5, simm13_lsb0:$Imm12),
+                                  "cv.beqimm", "$rs1, $Imm5, $Imm12", []>,
+                  Sched<[]>;
+  def CV_BNEIMM : RVInstImmBranch<0b011, (outs), (ins GPR:$rs1, simm5:$Imm5, simm13_lsb0:$Imm12),
+                                  "cv.bneimm", "$rs1, $Imm5, $Imm12", []>,
+                  Sched<[]>;
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]:
```suggestion
  def CV_BEQIMM : RVInstImmBranch<0b010, (outs), (ins GPR:$rs1, simm5:$imm5, simm13_lsb0:$imm12),
                                  "cv.beqimm", "$rs1, $imm5, $imm12", []>,
                  Sched<[]>;
  def CV_BNEIMM : RVInstImmBranch<0b011, (outs), (ins GPR:$rs1, simm5:$imm5, simm13_lsb0:$imm12),
                                  "cv.bneimm", "$rs1, $imm5, $imm12", []>,
                  Sched<[]>;
```

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 9]
```diff
@@ -232,45 +348,45 @@ def : Pat<(muls GPR:$rs1, GPR:$rs2),
...
           (CV_MULHHS GPR:$rs1, GPR:$rs2)>;
-def : Pat<(sra (muls GPR:$rs1, GPR:$rs2), uimm5:$Is3),
-          (CV_MULSN GPR:$rs1, GPR:$rs2, uimm5:$Is3)>;
-def : Pat<(sra (mulhhs GPR:$rs1, GPR:$rs2), uimm5:$Is3),
-          (CV_MULHHSN GPR:$rs1, GPR:$rs2, uimm5:$Is3)>;
-def : Pat<(shiftRound (muls GPR:$rs1, GPR:$rs2), uimm5:$Is3),
-          (CV_MULSRN GPR:$rs1, GPR:$rs2, uimm5:$Is3)>;
-def : Pat<(shiftRound (mulhhs GPR:$rs1, GPR:$rs2), uimm5:$Is3),
-          (CV_MULHHSRN GPR:$rs1, GPR:$rs2, uimm5:$Is3)>;
+def : Pat<(sra (muls GPR:$rs1, GPR:$rs2), uimm5:$imm5),
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td:351]: Ah thanks, I missed this in my last review.

**Review by flip1995 — COMMENTED** [review]
Last 2 NITs, then this should be good to go :+1:

**Comment by flip1995** [comment]
Thanks! Please squash your commits and this should be ready to merge.

**Comment by flip1995** [comment]
@jeremybennett This is ready to get merged :rocket:

**Comment by jeremybennett** [comment]
Thanks for all the work on this. Good to merge.
